File: apps/recommendations-ai-service/src/main/java/com/commerce/recommendations/controller/GlobalExceptionHandler.java
package com.commerce.recommendations.controller;

import com.commerce.recommendations.service.RecommendationGenerationException;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestControllerAdvice;

import java.time.Instant;
import java.util.Map;

@RestControllerAdvice
public class GlobalExceptionHandler {

    @ExceptionHandler(RecommendationGenerationException.class)
    @ResponseStatus(HttpStatus.BAD_GATEWAY)
    public Map<String, Object> handleRecommendationGenerationException(RecommendationGenerationException ex) {
        return Map.of(
                "error", "RECOMMENDATION_FAILED",
                "message", ex.getMessage(),
                "timestamp", Instant.now().toString()
        );
    }
}

File: apps/orders-service/src/main/kotlin/com/commerce/orders/controller/GlobalExceptionHandler.kt
package com.commerce.orders.controller

import com.commerce.orders.service.OrderNotFoundException
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice
import java.time.Instant

data class ApiError(
    val status: Int,
    val error: String,
    val message: String,
    val timestamp: Instant = Instant.now()
)

@RestControllerAdvice
class GlobalExceptionHandler {

    @ExceptionHandler(OrderNotFoundException::class)
    fun handleNotFound(ex: OrderNotFoundException): ResponseEntity<ApiError> =
        ResponseEntity.status(HttpStatus.NOT_FOUND).body(
            ApiError(HttpStatus.NOT_FOUND.value(), "NOT_FOUND", ex.message ?: "Not found")
        )

    @ExceptionHandler(MethodArgumentNotValidException::class)
    fun handleValidation(ex: MethodArgumentNotValidException): ResponseEntity<ApiError> {
        val details = ex.bindingResult.fieldErrors.joinToString(", ") {
            "${it.field}: ${it.defaultMessage}"
        }

        return ResponseEntity.badRequest().body(
            ApiError(HttpStatus.BAD_REQUEST.value(), "VALIDATION_FAILED", details)
        )
    }
}
